File: Complex.java
public class Complex {

	double re;
	double im;

	public Complex(double re, double im) {
		this.re = re;
		this.im = im;
	}

	public String toString() {
		return this.re + " " + this.im;
	}

	public Complex add(Complex other) {
		return new Complex(this.re+other.re, this.im+other.im);
	}

	public static void main(String[] args) {
		Complex z1 = new Complex(1,2);
		Complex z2 = new Complex(3,4);
		System.out.println(z1+z2);
	}
}

File: week07/complex.py
from __future__ import annotations

class Complex:

    def __init__(self, re:float, im:float) -> None:
        self.re = re
        self.im = im
    
    def __str__(self) -> str:
        return f"{self.re} + {self.im}i"
    
    def add(self, other:Complex) -> Complex:
        """Adds two numbers in the form
        z1 = a+ib
        z2 = c+id
        and returns 
        (a+c) + i*(b+d)
        """
        return Complex(self.re + other.re, self.im + other.im)

    def __add__(self, other:Complex) -> Complex:
        return self.add(other)

    def substract(self, other:Complex) -> Complex:
        return Complex(self.re - other.re, self.im - other.im)
    
    def __sub__(self, other:Complex) -> Complex:
        return self.substract(other)
    
    def __eq__(self, other:Complex) -> bool:
        return ((self.re == other.re) and (self.im == other.im))

